File: imu_calib_matlab/src/validate/plot_validation_results.m
function handles = plot_validation_results(data, calib, validation)
%PLOT_VALIDATION_RESULTS Plot validation figures for either full or partial results.

handles = struct();
mode = getfield_or_default(getfield_or_default(validation, 'summary', struct()), 'mode', 'full');
figMain = figure('Name', 'IMU Calibration Validation', 'Color', 'w');
tiledlayout(2, 3, 'Padding', 'compact', 'TileSpacing', 'compact');

ax1 = nexttile; plot_static_gyro_panel(ax1, data, validation);
ax2 = nexttile; plot_static_acc_panel(ax2, data, calib, validation);
ax3 = nexttile; plot_gyro_runs_panel(ax3, validation);
ax4 = nexttile; plot_cg_panel(ax4, calib);
ax5 = nexttile; plot_acc_summary_panel(ax5, validation, mode);
ax6 = nexttile; plot_summary_panel(ax6, calib, validation, mode);

handles.main = figMain;
handles.main_axes = [ax1, ax2, ax3, ax4, ax5, ax6];

allan = get_nested_or_default(validation, {'analysis', 'allan'}, struct());
if isstruct(allan)
    handles = add_allan_figures(handles, allan);
end

tempModels = getfield_or_default(calib, 'temp', struct());
if isstruct(tempModels)
    handles = add_temperature_figures(handles, data, tempModels);
end
end

function plot_static_gyro_panel(ax, data, validation)
axes(ax); %#ok<LAXES>
staticVal = getfield_or_default(validation, 'static', struct());
if isfield(staticVal, 'gyro_debiased') && ~isempty(staticVal.gyro_debiased) && ...
        isfield(data, 'static') && isfield(data.static, 't')
    plot(data.static.t, staticVal.gyro_debiased(:, 1), 'r');
    hold on;
    plot(data.static.t, staticVal.gyro_debiased(:, 2), 'g');
    plot(data.static.t, staticVal.gyro_debiased(:, 3), 'b');
    grid on;
    xlabel('Time [s]');
    ylabel('Gyro after bias removal [rad/s]');
    title('Static Gyro After Bias Removal');
    legend({'x', 'y', 'z'}, 'Location', 'best');
else
    text(0.5, 0.5, 'No static gyro available', 'HorizontalAlignment', 'center');
    axis off;
    title('Static Gyro');
end
end

function plot_static_acc_panel(ax, data, calib, validation)
axes(ax); %#ok<LAXES>
if ~isfield(data, 'static') || ~isfield(data.static, 't') || ~isfield(data.static, 'acc') || isempty(data.static.acc)
    text(0.5, 0.5, 'No static accel available', 'HorizontalAlignment', 'center');
    axis off;
    title('Static Accel Norm');
    return;
end

staticVal = getfield_or_default(validation, 'static', struct());
accNormRaw = getfield_or_default(staticVal, 'acc_norm_raw', []);
if isempty(accNormRaw)
    accNormRaw = sqrt(sum(data.static.acc .^ 2, 2));
end
accNormCorr = getfield_or_default(staticVal, 'acc_norm', []);

plot(data.static.t, accNormRaw, '--', 'Color', [0.6 0.6 0.6], 'LineWidth', 0.9);
hold on;
if ~isempty(accNormCorr)
    plot(data.static.t, accNormCorr, 'k', 'LineWidth', 1.0);
end

g0Line = getfield_or_default(calib, 'gravity_magnitude', []);
if isempty(g0Line)
    g0Line = default_calib_options().acc_calibration.gravity_magnitude;
end
yline(g0Line, '--r', 'Reference |a|');

grid on;
xlabel('Time [s]');
ylabel('Accel norm [m/s^2]');
title('Static Accel Norm');
if ~isempty(accNormCorr)
    legend({'Raw', 'Corrected', 'Reference |a|'}, 'Location', 'best');
else
    legend({'Raw', 'Reference |a|'}, 'Location', 'best');
end
end

function plot_gyro_runs_panel(ax, validation)
axes(ax); %#ok<LAXES>
gyroRuns = getfield_any(validation, {'gyro_runs', 'gyroRuns'}, []);
if ~isempty(gyroRuns)
    beforeVals = [gyroRuns.residual_norm_before].';
    afterVals = [gyroRuns.residual_norm_after].';
    bar([beforeVals, afterVals]);
    grid on;
    ylabel('||dtheta error|| [rad]');
    title('Gyro dtheta Error Before/After');
    legend({'Before', 'After'}, 'Location', 'best');
else
    text(0.5, 0.5, 'No gyro runs available', 'HorizontalAlignment', 'center');
    axis off;
    title('Gyro dtheta Error');
end
end

function plot_cg_panel(ax, calib)
axes(ax); %#ok<LAXES>
if isfield(calib, 'Cg') && ~isempty(calib.Cg)
    imagesc(calib.Cg);
    axis equal tight;
    colorbar;
    title('Cg Heatmap');
    xlabel('Reference axis');
    ylabel('Measured axis');
    set(gca, 'XTick', 1:3, 'XTickLabel', {'x', 'y', 'z'});
    set(gca, 'YTick', 1:3, 'YTickLabel', {'x', 'y', 'z'});
    for r = 1:3
        for c = 1:3
            text(c, r, sprintf('%.4f', calib.Cg(r, c)), ...
                'HorizontalAlignment', 'center', 'Color', 'w', 'FontWeight', 'bold');
        end
    end
else
    text(0.5, 0.5, 'Cg unavailable in partial mode', 'HorizontalAlignment', 'center');
    axis off;
    title('Cg Heatmap');
end
end

function plot_acc_summary_panel(ax, validation, mode)
axes(ax); %#ok<LAXES>
summary = getfield_or_default(validation, 'summary', struct());
if strcmp(mode, 'full') && isfield(summary, 'static_acc_norm_mean_before') && isfield(summary, 'static_acc_norm_mean_after')
    bar([summary.static_acc_norm_mean_before, summary.static_acc_norm_mean_after; ...
         summary.static_acc_norm_std_before, summary.static_acc_norm_std_after]);
    grid on;
    set(gca, 'XTick', 1:2, 'XTickLabel', {'Mean |a|', 'Std |a|'});
    legend({'Before', 'After'}, 'Location', 'best');
    title('Static Accel Summary');
elseif isfield(summary, 'static_acc_norm_mean') && ~isempty(summary.static_acc_norm_mean)
    bar([summary.static_acc_norm_mean; summary.static_acc_norm_std]);
    grid on;
    set(gca, 'XTick', 1:2, 'XTickLabel', {'Mean |a|', 'Std |a|'});
    title('Static Accel Summary');
else
    text(0.5, 0.5, 'No accel summary available', 'HorizontalAlignment', 'center');
    axis off;
    title('Static Accel Summary');
end
end

function plot_summary_panel(ax, calib, validation, mode)
axes(ax); %#ok<LAXES>
summary = getfield_or_default(validation, 'summary', struct());
lines = {sprintf('mode = %s', mode), ...
         sprintf('available_blocks = %s', stringify_list(getfield_or_default(summary, 'available_blocks', {}))), ...
         sprintf('completed_modules = %s', stringify_list(getfield_or_default(summary, 'completed_modules', {}))), ...
         sprintf('temp_status = %s', char(string(getfield_or_default(summary, 'temperature_model_status', 'n/a')))), ...
         sprintf('allan_status = %s', char(string(getfield_or_default(summary, 'allan_status', 'n/a'))))};
if isfield(summary, 'static_gyro_rms_after_bias') && ~isempty(summary.static_gyro_rms_after_bias)
    v = summary.static_gyro_rms_after_bias(:);
    lines{end + 1} = sprintf('gyro RMS after bias = [%.3g %.3g %.3g]', v(1), v(2), v(3)); %#ok<AGROW>
end
if isfield(summary, 'Ca_rcond') && ~isempty(summary.Ca_rcond)
    lines{end + 1} = sprintf('rcond(Ca) = %.3e', summary.Ca_rcond); %#ok<AGROW>
end
if isfield(summary, 'Cg_rcond') && ~isempty(summary.Cg_rcond)
    lines{end + 1} = sprintf('rcond(Cg) = %.3e', summary.Cg_rcond); %#ok<AGROW>
end
axis off;
title('Summary');
y = 0.92;
for i = 1:numel(lines)
    text(0.05, y, lines{i}, 'Units', 'normalized');
    y = y - 0.12;
end
end

function handles = add_allan_figures(handles, allan)
if isfield(allan, 'gyro') && isstruct(allan.gyro) && isfield(allan.gyro, 'valid') && allan.gyro.valid
    figGyro = figure('Name', 'Gyro Allan Deviation', 'Color', 'w');
    loglog(allan.gyro.tau, allan.gyro.adev(:, 1), 'r');
    hold on;
    loglog(allan.gyro.tau, allan.gyro.adev(:, 2), 'g');
    loglog(allan.gyro.tau, allan.gyro.adev(:, 3), 'b');
    grid on;
    xlabel('\tau [s]');
    ylabel('Allan deviation');
    title('Gyro Allan Deviation');
    legend({'x', 'y', 'z'}, 'Location', 'best');
    handles.allan_gyro = figGyro;
end

if isfield(allan, 'acc') && isstruct(allan.acc) && isfield(allan.acc, 'valid') && allan.acc.valid
    figAcc = figure('Name', 'Accel Allan Deviation', 'Color', 'w');
    loglog(allan.acc.tau, allan.acc.adev(:, 1), 'r');
    hold on;
    loglog(allan.acc.tau, allan.acc.adev(:, 2), 'g');
    loglog(allan.acc.tau, allan.acc.adev(:, 3), 'b');
    grid on;
    xlabel('\tau [s]');
    ylabel('Allan deviation');
    title('Accel Allan Deviation');
    legend({'x', 'y', 'z'}, 'Location', 'best');
    handles.allan_acc = figAcc;
end
end

function handles = add_temperature_figures(handles, data, tempModels)
if ~isfield(data, 'static') || ~isfield(data.static, 'temp') || isempty(data.static.temp)
    return;
end

bgModel = getfield_or_default(tempModels, 'bgModel', getfield_or_default(tempModels, 'bg_model', struct()));
if isstruct(bgModel) && isfield(bgModel, 'valid') && bgModel.valid
    handles.temperature_gyro = add_temperature_figure(data, bgModel, 'bg');
end

baModel = getfield_or_default(tempModels, 'baModel', getfield_or_default(tempModels, 'ba_model', struct()));
if isstruct(baModel) && isfield(baModel, 'valid') && baModel.valid
    handles.temperature_acc = add_temperature_figure(data, baModel, 'ba');
end
end

function figTemp = add_temperature_figure(data, model, target)
if strcmp(target, 'bg')
    [biasT, ~] = get_gyro_bias_from_temperature(data.static.temp(:), model.reference_bg, model);
    values = data.static.gyro;
    figName = 'Gyro Temperature Bias Model';
    prefix = 'bg';
    yunit = '[rad/s]';
else
    [biasT, ~] = get_accel_bias_from_temperature(data.static.temp(:), model.reference_ba, model);
    values = data.static.acc;
    figName = 'Accel Temperature Bias Model';
    prefix = 'ba';
    yunit = '[m/s^2]';
end

figTemp = figure('Name', figName, 'Color', 'w');
for axisIdx = 1:3
    subplot(3, 1, axisIdx);
    scatter(data.static.temp(:), values(:, axisIdx), 8, '.');
    hold on;
    [tempSorted, idx] = sort(data.static.temp(:));
    plot(tempSorted, biasT(idx, axisIdx), 'LineWidth', 1.2);
    grid on;
    xlabel('Temperature');
    ylabel(sprintf('%s_%d %s', prefix, axisIdx, yunit));
    title(sprintf('%s(T) axis %d', prefix, axisIdx));
end
end

function value = getfield_or_default(S, fieldName, defaultValue)
if isstruct(S) && isfield(S, fieldName)
    value = S.(fieldName);
else
    value = defaultValue;
end
end

function value = getfield_any(S, names, defaultValue)
value = defaultValue;
if ~isstruct(S)
    return;
end
for i = 1:numel(names)
    if isfield(S, names{i})
        value = S.(names{i});
        return;
    end
end
end

function value = get_nested_or_default(S, pathParts, defaultValue)
value = defaultValue;
current = S;
for i = 1:numel(pathParts)
    if ~isstruct(current) || ~isfield(current, pathParts{i})
        return;
    end
    current = current.(pathParts{i});
end
value = current;
end

function text = stringify_list(value)
if iscell(value)
    if isempty(value)
        text = '[]';
    else
        text = strjoin(cellfun(@(x) char(string(x)), value, 'UniformOutput', false), ', ');
    end
else
    text = char(string(value));
end
end

File: imu_calib_python/imu_calib/validate/plot_validation_results.py
from __future__ import annotations

import matplotlib.pyplot as plt
import numpy as np

from imu_calib.models.data_structures import ImuDataset


def plot_validation_results(data: ImuDataset, calib: dict, validation: dict) -> dict:
    """Plot validation figures for either full or partial results."""

    handles: dict[str, plt.Figure] = {}
    mode = validation.get("summary", {}).get("mode", "full")
    handles["main"] = _plot_main_figure(data, calib, validation, mode)

    allan = validation.get("analysis", {}).get("allan")
    if isinstance(allan, dict):
        handles.update(_add_allan_figures(allan))

    temp = calib.get("temp") if isinstance(calib, dict) else None
    if isinstance(temp, dict):
        temp_handles = _add_temperature_figures(data, temp)
        handles.update(temp_handles)
    return handles


def _plot_main_figure(data: ImuDataset, calib: dict, validation: dict, mode: str):
    fig, axes = plt.subplots(2, 3, figsize=(14, 8), num="IMU Calibration Validation")
    ax1, ax2, ax3, ax4, ax5, ax6 = axes.flatten()
    static = validation.get("static", {})
    summary = validation.get("summary", {})

    if data.static is not None and static.get("gyro_debiased") is not None:
        ax1.plot(data.static.t, static["gyro_debiased"][:, 0], "r", label="x")
        ax1.plot(data.static.t, static["gyro_debiased"][:, 1], "g", label="y")
        ax1.plot(data.static.t, static["gyro_debiased"][:, 2], "b", label="z")
        ax1.set_title("Static Gyro After Bias Removal")
        ax1.set_xlabel("Time [s]")
        ax1.set_ylabel("Gyro after bias removal [rad/s]")
        ax1.grid(True)
        ax1.legend()
    else:
        ax1.text(0.5, 0.5, "No static gyro available", ha="center", va="center")
        ax1.axis("off")

    acc_norm_raw = static.get("acc_norm_raw")
    acc_norm_corr = static.get("acc_norm")
    if data.static is not None and acc_norm_raw is not None:
        ax2.plot(data.static.t, acc_norm_raw, "--", color="0.6", label="Raw")
        if acc_norm_corr is not None:
            ax2.plot(data.static.t, acc_norm_corr, "k", label="Corrected")
        g0_line = _resolve_gravity_reference(data, calib, validation)
        if g0_line is not None:
            ax2.axhline(g0_line, color="r", linestyle="--", label="Reference |a|")
        ax2.set_title("Static Accel Norm")
        ax2.set_xlabel("Time [s]")
        ax2.set_ylabel("Accel norm [m/s^2]")
        ax2.grid(True)
        ax2.legend()
    else:
        ax2.text(0.5, 0.5, "No static accel available", ha="center", va="center")
        ax2.axis("off")

    gyro_runs = validation.get("gyro_runs", [])
    if gyro_runs:
        before = np.array([item["residual_norm_before"] for item in gyro_runs])
        after = np.array([item["residual_norm_after"] for item in gyro_runs])
        x = np.arange(before.size)
        width = 0.38
        ax3.bar(x - width / 2, before, width, label="Before")
        ax3.bar(x + width / 2, after, width, label="After")
        ax3.set_title("Gyro dtheta Error Before/After")
        ax3.set_ylabel("||dtheta error|| [rad]")
        ax3.grid(True)
        ax3.legend()
    else:
        ax3.text(0.5, 0.5, "No gyro runs available", ha="center", va="center")
        ax3.axis("off")

    if calib.get("Cg") is not None:
        im = ax4.imshow(calib["Cg"], aspect="equal")
        ax4.set_title("Cg Heatmap")
        ax4.set_xlabel("Reference axis")
        ax4.set_ylabel("Measured axis")
        ax4.set_xticks([0, 1, 2], ["x", "y", "z"])
        ax4.set_yticks([0, 1, 2], ["x", "y", "z"])
        for r in range(3):
            for c in range(3):
                ax4.text(c, r, f"{calib['Cg'][r, c]:.4f}", ha="center", va="center", color="w", fontweight="bold")
        fig.colorbar(im, ax=ax4, fraction=0.046, pad=0.04)
    else:
        ax4.text(0.5, 0.5, "Cg unavailable in partial mode", ha="center", va="center")
        ax4.axis("off")

    if mode == "full" and "static_acc_norm_mean_before" in summary and "static_acc_norm_mean_after" in summary:
        ax5.bar(
            np.arange(2),
            [summary["static_acc_norm_mean_before"], summary["static_acc_norm_std_before"]],
            width=0.35,
            label="Before",
        )
        ax5.bar(
            np.arange(2) + 0.35,
            [summary["static_acc_norm_mean_after"], summary["static_acc_norm_std_after"]],
            width=0.35,
            label="After",
        )
        ax5.set_xticks([0.175, 1.175], ["Mean |a|", "Std |a|"])
        ax5.set_title("Static Accel Summary")
        ax5.grid(True)
        ax5.legend()
    elif summary.get("static_acc_norm_mean") is not None:
        ax5.bar([0, 1], [summary["static_acc_norm_mean"], summary["static_acc_norm_std"]], width=0.5)
        ax5.set_xticks([0, 1], ["Mean |a|", "Std |a|"])
        ax5.set_title("Static Accel Summary")
        ax5.grid(True)
    else:
        ax5.text(0.5, 0.5, "No accel summary available", ha="center", va="center")
        ax5.axis("off")

    ax6.axis("off")
    ax6.set_title("Summary")
    lines = [
        f"mode = {summary.get('mode', mode)}",
        f"available_blocks = {summary.get('available_blocks', [])}",
        f"completed_modules = {summary.get('completed_modules', [])}",
        f"temp_status = {summary.get('temperature_model_status', 'n/a')}",
        f"allan_status = {summary.get('allan_status', 'n/a')}",
    ]
    if summary.get("static_gyro_rms_after_bias") is not None:
        vec = np.asarray(summary["static_gyro_rms_after_bias"], dtype=float).reshape(-1)
        lines.append(f"gyro RMS after bias = [{vec[0]:.3g} {vec[1]:.3g} {vec[2]:.3g}]")
    if "Ca_rcond" in summary and summary.get("Ca_rcond") is not None:
        lines.append(f"rcond(Ca) = {summary['Ca_rcond']:.3e}")
    if "Cg_rcond" in summary and summary.get("Cg_rcond") is not None:
        lines.append(f"rcond(Cg) = {summary['Cg_rcond']:.3e}")
    y = 0.92
    for line in lines:
        ax6.text(0.05, y, line, transform=ax6.transAxes)
        y -= 0.13

    fig.tight_layout()
    return fig


def _resolve_gravity_reference(data: ImuDataset, calib: dict, validation: dict) -> float | None:
    """Resolve the accel norm reference without requiring legacy a_ref fields."""
    static = validation.get("static", {}) if isinstance(validation, dict) else {}
    candidates = [static.get("gravity_magnitude")]
    if isinstance(calib, dict):
        candidates.append(calib.get("gravity_magnitude"))
        acc = calib.get("acc")
        if isinstance(acc, dict):
            candidates.append(acc.get("gravity_magnitude"))
    for value in candidates:
        if value is None:
            continue
        arr = np.asarray(value, dtype=float).reshape(-1)
        if arr.size == 1 and np.isfinite(arr[0]):
            return float(arr[0])

    legacy_refs = [
        float(np.linalg.norm(np.asarray(p.a_ref, dtype=float).reshape(3)))
        for p in data.acc_poses
        if getattr(p, "a_ref", None) is not None
    ]
    if legacy_refs:
        return float(np.mean(legacy_refs))
    return None


def _add_allan_figures(allan: dict) -> dict:
    handles: dict[str, plt.Figure] = {}
    if isinstance(allan.get("gyro"), dict) and allan["gyro"].get("valid", False):
        fig = plt.figure("Gyro Allan Deviation", figsize=(7, 5))
        for axis, color in enumerate(("r", "g", "b")):
            plt.loglog(allan["gyro"]["tau"], allan["gyro"]["adev"][:, axis], color=color, label=["x", "y", "z"][axis])
        plt.grid(True)
        plt.xlabel("tau [s]")
        plt.ylabel("Allan deviation")
        plt.title("Gyro Allan Deviation")
        plt.legend()
        handles["allan_gyro"] = fig
    if isinstance(allan.get("acc"), dict) and allan["acc"].get("valid", False):
        fig = plt.figure("Accel Allan Deviation", figsize=(7, 5))
        for axis, color in enumerate(("r", "g", "b")):
            plt.loglog(allan["acc"]["tau"], allan["acc"]["adev"][:, axis], color=color, label=["x", "y", "z"][axis])
        plt.grid(True)
        plt.xlabel("tau [s]")
        plt.ylabel("Allan deviation")
        plt.title("Accel Allan Deviation")
        plt.legend()
        handles["allan_acc"] = fig
    return handles


def _add_temperature_figures(data: ImuDataset, temp_models: dict) -> dict:
    handles: dict[str, plt.Figure] = {}
    if data.static is None or data.static.temp is None:
        return handles
    bg_model = temp_models.get("bg_model") if isinstance(temp_models, dict) else None
    ba_model = temp_models.get("ba_model") if isinstance(temp_models, dict) else None
    if isinstance(bg_model, dict) and bg_model.get("valid", False):
        fig = _add_temperature_figure(data, bg_model, target="bg")
        if fig is not None:
            handles["temperature_gyro"] = fig
    if isinstance(ba_model, dict) and ba_model.get("valid", False):
        fig = _add_temperature_figure(data, ba_model, target="ba")
        if fig is not None:
            handles["temperature_acc"] = fig
    return handles


def _add_temperature_figure(data: ImuDataset, model: dict, *, target: str):
    from imu_calib.runtime.get_gyro_bias_from_temperature import get_gyro_bias_from_temperature
    from imu_calib.runtime.get_accel_bias_from_temperature import get_accel_bias_from_temperature

    if target == "bg":
        biasT, _ = get_gyro_bias_from_temperature(data.static.temp, model["reference_bg"], model)
        values = data.static.gyro
        title = "Gyro Temperature Bias Model"
        yunit = "[rad/s]"
        prefix = "bg"
    else:
        biasT, _ = get_accel_bias_from_temperature(data.static.temp, model["reference_ba"], model)
        values = data.static.acc
        title = "Accel Temperature Bias Model"
        yunit = "[m/s^2]"
        prefix = "ba"
    fig, axes = plt.subplots(3, 1, figsize=(7, 8), num=title)
    temp = np.asarray(data.static.temp, dtype=float)
    order = np.argsort(temp)
    temp_sorted = temp[order]
    for axis_idx, ax in enumerate(axes):
        ax.scatter(temp, values[:, axis_idx], s=8, marker=".")
        ax.plot(temp_sorted, biasT[order, axis_idx], linewidth=1.2)
        ax.grid(True)
        ax.set_xlabel("Temperature")
        ax.set_ylabel(f"{prefix}_{axis_idx + 1} {yunit}")
        ax.set_title(f"{prefix}(T) axis {axis_idx + 1}")
    fig.tight_layout()
    return fig
